Swagger definition for Systems Management Service

chotea opened the pull request:
This is a refactoring for the current Systems Management HTTP API. We are writing a Service in C# so we can now define a Swagger. More details on what is changing can be found here: https://nitalk.jiveon.com/docs/DOC-500988

Signed-off-by: Cristian Hotea <cristian.hotea@ni.com>

chotea committed:
Swagger definition for Systems Management Service

This is a refactoring for the current SM Service. More details
on what is changing can be found here: https://nitalk.jiveon.com/docs/DOC-500988

Signed-off-by: Cristian Hotea <cristian.hotea@ni.com>

BKnight760 commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2030 @@
+        type: array
+        items:
+          type: string
+          example: invalidquery
```
'invalidquery' should not be in the args array.  The args array is for message inserts.

BKnight760 commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2030 @@
+        type: array
+        description: Internal errors that occured in the service.
+        items:
+          $ref: '#/definitions/Error'
```
For some reason, the example ends up as [ null ] which is not what we want.

chotea replied:
This is similar with other Swagger definitions. It seems that it renders the array, the description and inside the array is an empty dict.

BKnight760 commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2030 @@
+      jid:
+        description: The job ID of the job to cancel.
+        type: string
+        example: 20190211135234448026_28320
```
20190211135234448026_28320 gets rendered in scientific notation as "jid": 2.019021113523445e+24 which is probably not what was intended.  This happens in multiple places with this same value.

BKnight760 commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2030 @@
+              type: string
+      responses:
+        200:
+          $ref: '#/responses/Part
```
Should this route be just /v1/query-jobs?  That seems to be the preferred style in other apis.

BKnight760 commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2030 @@
+        description: The time when the system has been registered.
+        type: string
+        format: iso-date-time
+        example: 2019-02-19T11:42:25.078Z
```
The example for all of the timesamp elements in this definition render as an empty object: {} instead of an ISO-8601 string value.

BKnight760 commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2030 @@
+        description: Properties metadata information about a system.
+        type: object
+        properties:
+          last_updated_timestamp:
```
This 'last_updated_timestamp' needs an example element so that it does not render as "string".  This is needs to be addressed in multiple places in this document.

BKnight760 commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2030 @@
+        description: Information for how to authenticate.
+        type: string
+
+  PartialSuccess:
```
This partial success response only includes an error object.  What data is included for the items that were successful?  Generally, the partial success response definition needs to be specific to each type of resource / route.

chotea replied:
For almost all routes I think returning the failed items is enough. The only route that could be addressed is /v1/create-groups which could return the generated gids. I will modify it.

BKnight760 replied:
The "Responses" section of the SystemLink HTTP API Guidelines indicates 'partial success' responses should contain more than just the error definition.

https://nitalk.jiveon.com/docs/DOC-410194

chotea replied:
That is right: "If any resources were created, should include a success array of objects in the response". So in this case only the CreateGroups route should contain additional information. The other routes are RPCs so I think they are fine.

BKnight760 commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2030 @@
+      ip_address:
+        type: string
+        description: Required if hostname is not provided. A string representing the IP address of the system to register.
+        example: 1.1.1.1
```
Other places use 10.10.10.10 as the example ip_address.  I personally like that better than 1.1.1.1.

BKnight760 commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2030 @@
+          $ref: '#/responses/Unauthorized'
+        default:
+          $ref: '#/responses/Error'
+  /v1/get-systems-report:
```
I'd advise against using 'get' in routes names that are HTTP POST verbs instead of HTTP GET to avoid confusion.

BKnight760 commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2030 @@
+    properties:
+      type:
+        $ref: '#/definitions/ReportType'
+      ids:
```
Does this need to be an array of IDs, or could this use the new query syntax as a filter for any matching systems, including a list of IDs?

chotea replied:
good catch!

BKnight760 commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2030 @@
+          $ref: '#/responses/Unauthorized'
+        default:
+          $ref: '#/responses/Error'
+    get:
```
This is odd to have a GET verb for /create-groups.  It seems that there should be a /groups resource on which you would do a POST or GET.

chotea replied:
I've splitted the initial /groups route into 2 separate routes and I forgot to add the URL for the get route. I can't use a POST to /groups for CreateGroups because we are creating multiple resources, so is RPC instead of REST.

BKnight760 commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2030 @@
+          schema:
+            type: array
+            items:
+              type: string
```
This needs an example, otherwise it renders as ["string"]

BKnight760 reviewed (dismissed):
There are several items that are easily identifiable by opening each route in the swagger editor and viewing the example request and response bodies.  A few routes should be evaluated for naming and usage based on the HTTP verbs that are being used.

alexweav commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2030 @@
+          $ref: '#/responses/Unauthorized'
+        default:
+          $ref: '#/responses/Error'
+  /v1/get-systems-report:
```
The old API had a way of automatically uploading the systems software report to the file ingestion service and returning a file ID instead of the actual report contents. Do we plan to continue supporting this use case in the new API?

chotea replied:
Only the current AMQP API supports that so we will keep this functionality only in AMQP API.

chotea committed:
Implemented feedback

Signed-off-by: Cristian Hotea <cristian.hotea@ni.com>

chotea committed:
We will support only multi function jobs.

Changed the JobModel to support only multi-function requests. Fun has
to be a list of strings. We don't support 'string' case anymore because
we want to be able to use DyanmicExpressions to query for jobs data.
They don't work if the JobModel is not "well" defined(inheritances are not supported)
So to be able to use the projection over the models in
backend we will support only arrays. And the API looks much better now.

Signed-off-by: Cristian Hotea <cristian.hotea@ni.com>

skizunov commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+        description: The job ID of the job to cancel.
+        type: string
+        example: '20190211135234448026_28320'
+      systemId:
```
You have `request_id` above and `systemId` here. Is there a reason why some use lowercase_with_underscores and others use camelCase?

chotea replied:
I will make all properties camel cased.

skizunov commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+    type: object
+    properties:
+      connected_count:
+        type: integer
```
For Python based codegen (both service and client-side), each time you use `type: integer`, it also needs either `format: int32` or `format: int64`. I would suggest you go through and add these to each and every use of `type: integer` that you have.

If you think about it, it is a silly requirement since in Python 3 everything is just an `int`. But it is a current requirement. I personally tend to lean towards `int64` for most of these, so that any potential C++ codegen makes sense (since in C++, they tend to use size_t for most things, which is 64-bit on 64-bit machines).

skizunov commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+        description: Specifies the filter criteria for resources.
+      projection:
+        type: string
+        example: 'new(id,grains,last_updated_timestamp)'
```
Curious, why isn't this an array of strings or enums? Also, in your example, what does the `new` at the front mean (with parenthesis)?

chotea replied:
For now we don't want to restrict the API to certain keys when using filtering and projection(and I don't think we can do that since we have the grains dictionary that can contain anything). So we let the user to send a string that should be built based on the JobModel/SystemModel properties. We will make sure that the models aren't changed so we will avoid breaking the API. The "new" operator implies creation of a new object(dict). Without that, the query will return a tuple with the values instead of a dict. So the "new" operator is very important when you want to keep the hierarchy of the response and to avoid flattening. This way the client doesn't have to write custom logic to handle the query response for each different query request.

skizunov commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+      take:
+        description: How many resources to take. Maximum value is 1000.
+        type: integer
+        default: -1
```
Instead of defaulting to -1, you can make this optional (default to null) and then have the minimum be 0 (eg by using `minimum: 0`). This way you don't have -1 mean a magic value, whereas the other numbers have a different meaning.

skizunov commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+        description: How many resources to skip.
+        type: integer
+        default: 0
+        example: 100
```
You could add `minimum: 0` here.

skizunov commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+              example: 'http://download.ni.com/support/nipkg/products/ni-package-manager/released'
+      responses:
+        200:
+          $ref: '#/responses/PartialSuccess'
```
The `PartialSuccess` object does not contain the feeds that were successfully configured nor the feeds that failed.

chotea replied:
The PartialSuccess response will contain the information about the feeds that were not configured(innerErrors). I don't think we should add any additional information because:
"If any resources were created, should include a success array of objects in the response"
We are not managing feeds with this API, we are not assigning IDs to them, so I don't see them as resources in this API. Our high-level use case is to get the available packages from a list of feeds. Ideally this route and the getAvailablePackages should be merged into one single route but we decided to keep them separated(to leverage the current implementations from SMS) and merge them only if we have enough time before the release.

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+﻿swagger: '2.0'
+info:
+  description: 'Management of distributed systems.'
```
"Systems Management Service HTTP API. Manage and monitor your distributed systems."

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+        type: string
+
+  PartialSuccess:
+    description: Partial Success in the case when only a part of the request body metadata information was processed successfully.
```
"Partial Success indicates only part of the request body metadata information was processed successfully."?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+
+definitions:
+  CreateJobRequest:
+    description: Request describing a job to be sent to systems.
```
"Request describing a job to send to systems."

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+          - pkg.remove
+          - pkg.install
+      arg: 
+        description: An array of arrays of variable type elements that are arguments to the function specified by the "fun" property. If at least one function requires arguments, an array must be provided for each function, even if it is empty. Keyword arguments may be provided as a parameter of type object with at least one property of type boolean, "__kwarg__", that is evaluated to true.
```
"An array of variable type element arrays, which are arguments to the function specified by the "fun" property."?

caleyburton replied:
"... boolean ("_ kwarg _") that evaluates to true."? Sorry, I had to add spaces between the underscores and "kwarg" because markdown removes the underscores and italicize the "kwarg" when they're together.

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+              pkgs:
+              - acl-dbg: 1.0.0-0
+      metadata:
+        description: Additional information passed along with the job.
```
"passed to the system with the job" maybe?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+
+  CreateJobResponse:
+    title: Create Job Response
+    description: Response describing a job that was sent to systems.
```
"Response describing a job sent to systems."?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+    required: [systems_credentials]
+
+  SystemCredentials:
+    description: Credentials of a system.
```
"The hostname, IP address, and login information of a system."?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+
+  RegisterSystemsResponse:
+    title: Register Systems Response
+    description: Response of a request to register systems.
```
"Response to a request to register systems to the server." or  "Response to a request to register systems to the Systems Management service."?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+
+  UnregisterSystemsRequest:
+    title: Unregister Systems Request
+    description: Request to unregister systems.
```
Would it be useful to say "Request to unregister systems from the server" or "Request to unregister system from the Systems Management service"?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+    type: object
+    properties:
+      tgt:
+        description: Array of string representing the IDs of systems to be removed.
```
"Array of strings representing the IDs of systems to remove from SystemLink Server." or "Array of strings representing the IDs of systems to remove from the Systems Management service."?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+          type: string
+          example: VMware_Virtual_Platform--SN-VMware-56_4d_30_26_4f_c9_5c_05-b7_68_d1_cd_99_6e_22_16--MAC-02-01-02-03-04-08
+      force:
+        description: Specify as true to remove systems from the database immediately, or false to wait until the unregister job returns from systems. If true, unregister job failures will not be cached.
```
"A Boolean which specifies whether to remove systems from the database immediately (True) or wait until the unregister job returns from systems (False). If True, unregister job failures are not cached."?

chotea committed:
Implemented feedback from Sergey

Signed-off-by: Cristian Hotea <cristian.hotea@ni.com>

chotea committed:
Removed x-ni-data-type usings

Signed-off-by: Cristian Hotea <cristian.hotea@ni.com>

skizunov commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2093 @@
+        type: integer
+        format: int32
+        description: Whether the package is essential or not. A value greater than 0 means the package is essential.
+        example: 1
```
Is there a reason why this isn't a `boolean` instead of an `integer`? Based on the description, it sounds like a bool to me.

chotea replied:
NIPM C API gives us a numeric value. You can check it here: Perforce\AST\PackageManagement\components\nipkg\export\19.0\19.0.0f0\includes\nipkg_client\shared\constants.h 
The current implementation uses an integer and I thought I can leave it this way. But looking into C# API of NIPM I see that they made it a boolean. So after all it seems it should have been a boolean. I will change it.

chotea committed:
Made essential a boolean.

Signed-off-by: Cristian Hotea <cristian.hotea@ni.com>

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+          title: Failed Unregister Request
+          properties:
+            id:
+              description: The ID of the system on which unregister operation failed.
```
"unregister the operation"

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+              $ref: '#/definitions/Error'
+
+  CancelJobRequest:
+    description: Object containing the job ID and a system ID.
```
Could we remove "a" ("the job ID and system ID")?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+        type: string
+        example: '20190211135234448026_28320'
+      systemId:
+        description: The system ID on which the job will be canceled.
```
"The system ID on which the job is canceled."?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+    properties:
+      discovered_count:
+        type: integer
+        description: Number of systems that are discovered and not managed.
```
"not managed by your SystemLink Server" or "not managed by the Systems Management service"?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+
+  PendingSystemsSummary:
+    title: Pending Systems Summary
+    description: Summary information about the systems that are trying to connect to the server but aren't accepted yet.
```
"Summary information about the systems attempting to connect to the server but aren't accepted yet."

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+    properties:
+      pending_count:
+        type: integer
+        description: Number of systems that are currently pending to be added to the master but have not yet been approved by the master.
```
"Number of systems pending approval from the master to connect to the master."?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+        type: string
+        example: Test system
+      cpu_arch:
+        description: System CPU architecture
```
Could we add a period here?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+    description: > 
+      The state of the job.
+       * Canceled: The job was canceled by the user.
+       * Failed: The job has completed with failures. The job is failed if at least one of its functions being executed has failed.
```
"The job completed with failures. The job fails if at least one of the its functions executing fails."?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+      The state of the job.
+       * Canceled: The job was canceled by the user.
+       * Failed: The job has completed with failures. The job is failed if at least one of its functions being executed has failed.
+       * Succeeded: The job has completed with success.
```
"The job completed successfully."

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+       * Canceled: The job was canceled by the user.
+       * Failed: The job has completed with failures. The job is failed if at least one of its functions being executed has failed.
+       * Succeeded: The job has completed with success.
+       * InQueue: The job is in queue waiting to be dispatched to the system.
```
Could we add a comma between "queue" and "waiting" (i.e. "queue, waiting")?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+            items:
+              type: boolean
+              example: false
+            description: An array of booleans representing succees values for each executed function.
```
"success"

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+       * Approved: The system is approved but it didn't try to connect yet.
+       * Disconnected: The system is disconnected.
+       * ConnectedRefreshPending: The system is connected and the refresh job is running. A refresh job is sent to the system every time the connection is established in order to gather the system information.
+       * Connected: The system is connected. The refresh job has completed with success.
```
"The refresh job completed successfully."

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+       * Disconnected: The system is disconnected.
+       * ConnectedRefreshPending: The system is connected and the refresh job is running. A refresh job is sent to the system every time the connection is established in order to gather the system information.
+       * Connected: The system is connected. The refresh job has completed with success.
+       * ConnectedRefreshFailed: The system is connected but the refresh job has failed.
```
"The system is connected but the refresh job failed."

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+       * ConnectedRefreshFailed: The system is connected but the refresh job has failed.
+
+  ManagedSystemActivationData:
+    description: Information about the license of a system
```
Could we add a period at the end?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+        description: The ID of the system.
+        example: VMware_Virtual_Platform--SN-VMware-56_4d_d2_08_ce_89_1b_70-e6_32_fc_21_6c_d0_d9_f9--MAC-00-0C-29-D0-D9-F9
+      created_timestamp:
+        description: The time when the system has been registered.
```
"was registered" instead?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+        format: iso-date-time
+        example: '2019-02-19T11:42:25.078Z'
+      last_updated_timestamp:
+        description: Last time when the system information has been updated.
```
"updated" instead of "has been updated"?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+        type: object
+        properties:
+          last_updated_timestamp:
+            description: Last time when the system activation information has been updated.
```
"updated" instead of "has been updated"?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+            osversion: 6.1.7601
+            osrelease: 7        
+      packages:
+        description: Software packages that are installed on the system.
```
remove "that are"?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+              group: Infrastructure
+              version: 19.5.0.28-0+d28
+      feeds:
+        description: Feeds that are configured on the system.
```
remove "that are"?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+                tag: 16781312
+                value: false
+      groups:
+        description: Groups information about a system. Specifies on which groups is the system assigned to.
```
"Specifies the groups to which the system is assigned."

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+    properties:
+      groups:
+        type: array
+        description: The groups that got created.
```
"The groups created."?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+    properties:
+      systems_pending:
+        type: object
+        description: An object of system ID to key key/value pairs that are pending approval or rejection. Keys may be pending when they are not pre-seeded via the automated registration process and attempt to connect to the master.
```
Is there another word we can use instead of "key" here? I ask this because "to key key/value pairs" might confuse some customers. Would "locate" or "find" work here? This comment applies to all of the following descriptions with "key key/value pairs" in it.

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+        required: [key_actions]
+
+  SystemPatchRequest:
+    description: System data that has to be assigned to the system.
```
could we say "needs" instead of "has" here?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+        description: An array containing information about other packages this package breaks.
+        items:
+          type: string
+          description: The name of a package that this package breaks. This also contains information about the version of the package.
```
could we remove "that" (i.e. "that this package breaks" --> "package this package breaks")? This removal request applies to the following descriptions that have "that this" in them.

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+        example: "ni-securityupdate-kb67l8lcqw-killbits"
+      plugin:
+        type: integer
+        description: Agent that will be used to install the Package.
```
could we lowercase "package"?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+        example: "thunar-archive-plugin_0.3.1.bb"
+      storeproduct:
+        type: boolean
+        description: Whether the package represent a product in the Store or not.
```
Could we lowercase "store"? Furthermore, should we specific "the NI store"?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+        description: >-
+          Available operations in the v1 version of the API:
+
+          - getJobs: The ability to retrieve the jobs
```
Could we add periods to all of the descriptions here?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+          
+          - manageFeeds: The ability to configure feeds and retrieve the available packages
+          
+          - createOrUpdateGroups: The ability to create, update or delete groups
```
"The ability to create, update, or delete groups."

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+
+          - getGroups: The ability to retrieve the groups
+    
+          - manageSystems: The ability to register/unregister systems
```
Let's say "register or unregister" here?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+        default: 0
+        example: 100
+      take:
+        description: How many resources to take. Maximum value is 1000.
```
"The number of resources to return. The maximum value is 1000."

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+    title: Query Request
+    properties:
+      skip:
+        description: How many resources to skip.
```
"The number of resources to skip."

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+      filter:
+        type: string
+        example: connected.data.state==Connected
+        description: Specifies the filter criteria for resources.
```
"The criteria used to filter resources."?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+      orderBy:
+        type: string
+        example: 'last_updated_timestamp descending'
+        description: Specifies the ordering of returned data.
```
"The order in which data returns."?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+      projection:
+        type: string
+        example: 'new(id,grains,last_updated_timestamp)'
+        description: Specifies the projection for resources.
```
What does "projection" mean here? Are we creating the resource request with this parameter?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+        - in: body
+          name: CreateJobRequest
+          required: true
+          description: The definition of job to be executed
```
Could we add a period to then end?

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+      x-ni-operation: createOrCancelJobs
+      tags: [jobs]
+      summary: Cancel jobs
+      description: Cancel jobs that are running on systems. This won't stop the jobs execution on systems if the jobs have already been dispatched to the systems.
```
"This won't stop the jobs from executing on systems"

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+      x-ni-operation: getJobs
+      tags: [jobs]
+      summary: Get jobs summary
+      description: Get the number of active, failed and succeeded jobs.
```
"active, failed, and succeeded jobs"

caleyburton commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2076 @@
+          description: A group ID.
+      responses:
+        200:
+          descript
```
Missing a period.

chotea committed:
Implemented feedback from Caley

Signed-off-by: Cristian Hotea <cristian.hotea@ni.com>

chotea commented:
> There are several items that are easily identifiable by opening each route in the swagger editor and viewing the example request and response bodies. A few routes should be evaluated for naming and usage based on the HTTP verbs that are being used.

vdeac commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2092 @@
+      filter:
+        type: string
+        example: connected.data.state==CONNECTED
+        description: The criteria used to filter resources.
```
For the filter, I recommend that our public APIs are explicit and describe both the syntax of the language and explicitly state the properties which can be used in the filter.

Take a look at the [Query-Assets](http://demo.systemlink.io/niapis/?urls.primaryName=Asset%20Performance%20Management#/assets/query-assets ) request and you can use the same text here too.

vdeac commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2092 @@
+      x-ni-operation: createOrCancelJobs
+      tags: [jobs]
+      summary: Create a job
+      description: Create a job.
```
Would it make sense to also mention what is the effect of the route (creates a job and returns the newly created job)?

vdeac commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2092 @@
+          schema:
+            $ref: '#/definitions/UnregisterSystemsRequest'
+      responses:
+        200:
```
Seems this could also return partial success, if only a subset of systems were unregistered.

vdeac commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2092 @@
+      tags: [systems]
+      summary: Get discovered systems
+      description: Get network discovered systems.
+      responses:
```
Doesn't this route need a 401 and default responses?

vdeac commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2092 @@
+          $ref: '#/responses/Error'
+  /v1/groups:
+    get:
+      oper
```
It seems this would need a 404 Not Found response if no group matches the provided id

vdeac replied:
404 not found if no packages match the request parameters?

vdeac commented on systems-management/nisysmgmt.yml:
```diff
@@ -0,0 +1,2092 @@
+          $ref: '#/responses/Error'
+  /v1/groups:
+    get:
+      oper
```
404 Not found ?

chotea replied:
I don't feel that we should return a 404 in this case. The input is always an array of credentials, so having one element in that array is a particular case. Checking if the array contains only one element to return 404 in error case would complicate the handling of the response on the client side since the response of a request containing more than 1 element would come with 200 and a specific schema.

vdeac reviewed (dismissed):
Overall looks good.
I would advise to add more documentation around the filtering/projection parameters in the requests and also add missing responses to some of the routes.

chotea committed:
Implemented feedback from Vlad

Signed-off-by: Cristian Hotea <cristian.hotea@ni.com>

chotea committed:
Fixed feedUris required.

Signed-off-by: Cristian Hotea <cristian.hotea@ni.com>
